### PR TITLE
#170947676 Add Search Functionality

### PR DESCRIPTION
**Pull request by Niyonsengaeric**
#### What does this PR do?
- Add Search Functionality.
#### Description of Task to be completed?
- Check if the user is logged in.
- Check if the user has entered a search key.
- Search in the table if the search key match.
- Display the results.
- The manager can search in all his requesters Data.
- Test the endpoint.
- Document the endpoint using swagger documentation.
#### How should this be manually tested?
- run `git clone https://github.com/andela/knights-bn-backend.git` to clone the repo.
- run `cd knights-bn-backend` in your terminal to enter into the repository.
- run `git checkout ft-search-functionality-170947676` to switch to this branch.
- run `npm install` to install dependencies.
- create a Postgres database.
- create a .env file then add environment variable as shown below:
```
DB_USER='name of the user of the database'
DB_PASS='password of your Postgres'
DB_NAME='name of the database || Barefoot'
DB_TEST_NAME='name of the database for test || Barefoot_test'
NODE_ENV='development'
SECRETKEY='PrivateKey'
BN_API_KEY='use the SendGrid given ken'
```
- use .env.example as a blueprint for your .env file.
- run `npm run createTables` to create the database tables.
- run `npm dev-start` to start the server.
- open the browser use URL:`http://localhost:4000/api/v1/docs/#/Trips/SearchFunctionality` for swagger documentation.
- Open postman and copy this URL POST `http://localhost:4000/api/v1/auth/login`. to login to the system if you have an account.
- If you don't have an account open postman and copy this URL POST `localhost:4000/api/v1/auth/signup`  to sign up
- Use this format to Sign up
```
{
	"firstName": "Kagorora",
	"lastName": "Messi",
	"gender": "Male",
	"PassportNumber": "12345675",
	"email": "example@gmail.com ",
	"password": "Niyonkuru@1"
}
```
- Use this user to login as manager
```
{
        "email":"eugene.munyampundu@gmail.com",
	"password":"Niyonkuru@1"
}
             
```
- If a user is successfully signup or login 
#### Add Search Functionality
 - Open postman and copy and send this URL GET `localhost:4000/api/v1/trips/search?filterKey=your desire search key`.
          FilterKey must be either:` requestId, requesterId, type, status, reason, origin, destination`
- You will see the result according to your search key.
#### Any background context you want to provide?
N/A
#### What are the relevant pivotal tracker stories?
[#170947676](https://www.pivotaltracker.com/story/show/170947676)
#### Screenshots 
                 Sign up a new user 

Screen Shot 2020-03-01 at 23 23 57

                Search result for the keyword `California` for the new user

Screen Shot 2020-03-03 at 09 55 31

             Login as a Manager

Screen Shot 2020-03-01 at 23 23 57

             Search result for the keyword `Kigali`

Screen Shot 2020-03-03 at 09 52 35

#### Questions:
N/A

### CODE REVIEW

File: src/controllers/request.js [hunk 1]
```diff
@@ -95,4 +99,80 @@ export default class usersController {
...
+        ),
+        where(
+          cast(col('status'), 'varchar'),
+          { [Op.like]: `%${filterKey}%` }
+        ),
+        { reason: { [Op.like]: `%${filterKey}%`, }, },
+        { origin: { [Op.like]: `%${filterKey}%`, }, },
+        { destination: { [Op.like]: `%${filterKey}%`, }, },        
+      ]; 
+
```
Comment (Moise1) [inline src/controllers/request.js]: @Niyonsengaeric  please use the case insensitive feature for the best user experience. The user should be able to get results when they used either lower case or upper case search queries.

Comment (Niyonsengaeric) [reply src/controllers/request.js]: Done!

**Comment by Kagorora** [comment]
@Niyonsengaeric , I wanted to search request of id=1 but I got two(remember requestId is unique), any explanations

Screen Shot 2020-03-03 at 08 45 30

@Kagorora That means it filters all relevant information that contains that keyword `1`

**Comment by Niyonsengaeric** [comment]
> @Niyonsengaeric , I wanted to search request of id=1 but I got two(remember requestId is unique), any explanations
> 
> Screen Shot 2020-03-03 at 08 45 30
> 
> @Kagorora That means it filters all relevant information that contains that keyword `1`

@Kagorora you can now check it again I have added the precision of what you want to search

File: src/controllers/request.js (hunk 1 above)
Comment (Moise1) [reply src/controllers/request.js]: Good job @Niyonsengaeric .

**Review by Moise1 — APPROVED** [review]

**Review by Kagorora — APPROVED** [review]

**Review by kagabof — APPROVED** [review]
Great work.

**Pull request merged**